### PR TITLE
Add Region and Endpoint Configuration Options to AWS CloudProvider

### PR DESCRIPTION
**Pull request by BugRoger**
The aws-sdk-go supports custom endpoints and regions for EC2 compatible clouds. This change allows to configure a region and endpoint in the aws cloud provider config. These parameters are passed through to the aws client.

A complete `cloud-config` file would look like:

```
[global]
zone = eu-de-1
region = europe
endpoint = https://eu-de-1.europe.ec2.acmecorp.com
```

Coincidentally, this also leads to the region validation to fail if the EC2 compatible cloud is, well not, 100% compatible and doesn't adhere to the Amazons naming scheme. The assumption is that the region can be calculated by striping the index of the availability zone. Since this is not always the case, an explicit region in the config will not be validated.

### CODE REVIEW

**Comment by BugRoger** [comment]
rebased

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -57,9 +57,9 @@ const MaxReadThenCreateRetries = 30
 
 // Abstraction over AWS, to allow mocking/other implementations
 type AWSServices interface {
-	Compute(region string) (EC2, error)
-	LoadBalancing(region string) (ELB, error)
-	Autoscaling(region string) (ASG, error)
+	Compute(region string, endpoint string) (EC2, error)
+	LoadBalancing(region string, endpoint string) (ELB, error)
+	Autoscaling(region string, endpoint string) (ASG, error)
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: I'm almost wondering if we should just pass in *aws.Config.  For now it's fine, but I think if we get to 4 arguments we should switch!

File: pkg/cloudprovider/providers/aws/aws.go [hunk 2]
```diff
@@ -191,7 +192,9 @@ type AWSCloudConfig struct {
 	Global struct {
 		// TODO: Is there any use for this?  We can get it from the instance metadata service
 		// Maybe if we're not running on AWS, e.g. bootstrap; for now it is not very useful
-		Zone string
+		Zone     string
+		Region   string
+		Endpoint string
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Is there any way to get the endpoint from the metadata service on your AWS implementation?  And which one are you using (we might actually be able to test against it, which could be cheaper than testing against AWS!)

File: pkg/cloudprovider/providers/aws/aws.go [hunk 3]
```diff
@@ -554,19 +560,25 @@ func newAWSCloud(config io.Reader, awsServices AWSServices) (*AWSCloud, error) {
 	if len(zone) <= 1 {
 		return nil, fmt.Errorf("invalid AWS zone in config file: %s", zone)
 	}
-	regionName := zone[:len(zone)-1]
+
+	regionName := cfg.Global.Region
+	if len(regionName) <= 1 {
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Do you mean `regionName == ""` or `len(regionName) == 0`?  The len(zone) <=1 test above is because we're about to slice the last character off.

Maybe this logic is more correct?

```
regionName := cfg.Global.Region
  if regionName == "" {
    if len(zone) <= 1 {
      return nil, fmt.Errorf("invalid AWS zone in config file: %s", zone)
    }
    regionName = zone[:len(zone)-1]
}
```

File: pkg/cloudprovider/providers/aws/aws.go [hunk 4]
```diff
@@ -554,19 +560,25 @@ func newAWSCloud(config io.Reader, awsServices AWSServices) (*AWSCloud, error) {
...
+		regionName = zone[:len(zone)-1]
+	}
+
+	endpoint := cfg.Global.Endpoint
 
 	valid := isRegionValid(regionName)
-	if !valid {
-		return nil, fmt.Errorf("not a valid AWS zone (unknown region): %s", zone)
+	if len(cfg.Global.Region) <= 0 && !valid {
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: The `<=` is odd.  I think `len(s) == 0` or `s == ""`

But actually I think you only want to make this check if we computed regionName from the zone, so maybe move it into the `cfg.Global.Region == ""` block above?

File: pkg/cloudprovider/providers/aws/aws_test.go [hunk 5]
```diff
@@ -196,34 +216,51 @@ func TestNewAWSCloud(t *testing.T) {
...
 		},
 		{
 			"Config specified invalid zone",
 			strings.NewReader("[global]\nzone = blahonga"), NewFakeAWSServices(),
-			true, "",
+			true, "", "", "",
 		},
 		{
-			"Config specifies valid zone",
+			"Config specifies valid zone, calulate region",
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_test.go]: Typo (in at least one other place that I spotted): `s/calulate/calculate`

**Comment by justinsb** [comment]
Generally looks good, a few things on the logical flow around regions when _not_ using the config override.

File: pkg/cloudprovider/providers/aws/aws.go (hunk 1 above)
Comment (BugRoger) [inline pkg/cloudprovider/providers/aws/aws.go]: Agreed.

File: pkg/cloudprovider/providers/aws/aws.go (hunk 2 above)
Comment (BugRoger) [inline pkg/cloudprovider/providers/aws/aws.go]: Not in it's current implementation. The problem is that the `aws-sdk-go` calculates the endpoint from the region. We made the mistake to not follow the exact same naming scheme. It's virtually impossible to change it now, so we are stuck with configuring the endpoints for each client... :cry: 

Be careful what you wish for... :smile: We are running a home grown (so far) non-public AWS clone. The e2e test should stay on Amazon. Even the API of our cloud is compatible there are a lot of unspecced edge cases which behave slightly differently. 

I haven't looked into running the e2e tests against our cloud. My assumption was that the time to set it up would be quite prohibitive.

File: pkg/cloudprovider/providers/aws/aws.go (hunk 3 above)
Comment (BugRoger) [inline pkg/cloudprovider/providers/aws/aws.go]: Good suggestion. I was cargo culting. :sweat_smile:

File: pkg/cloudprovider/providers/aws/aws.go (hunk 4 above)
Comment (BugRoger) [inline pkg/cloudprovider/providers/aws/aws.go]: Yes, totally. Changed.

File: pkg/cloudprovider/providers/aws/aws_test.go (hunk 5 above)
Comment (BugRoger) [inline pkg/cloudprovider/providers/aws/aws_test.go]: Fixed.

**Comment by BugRoger** [comment]
Thanks for the review, @justinsb. :smile:

**Comment by justinsb** [comment]
Sorry this one fell off my radar.  It LGTM, though I'm afraid you'll have to rebase first.

**Comment by justinsb** [comment]
ok to test

**Comment by justinsb** [comment]
@BugRoger do you still want this?  If so you'll have to rebase...  Otherwise we can revisit it for 1.3

**Comment by BugRoger** [comment]
Thanks for the reminder. I'll rebase...

**Comment by fejta** [comment]
This PR has had no meaningful activity for multiple months. If it is still valid please rebase, push a new commit and reopen the PR. Thanks!
